File: customer-app/lib/services/wallet_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_service.dart';

final walletServiceProvider = Provider<WalletService>((ref) {
  final apiService = ref.watch(apiServiceProvider);
  return WalletService(apiService);
});

class WalletService {
  final ApiService _apiService;

  WalletService(this._apiService);

  /// Get wallet balance
  Future<WalletBalance> getBalance() async {
    try {
      final response = await _apiService.get('/customer/wallet/balance');

      if (response.data['success']) {
        return WalletBalance.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل جلب رصيد المحفظة');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء جلب رصيد المحفظة',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Get wallet transactions
  Future<WalletTransactionsResponse> getTransactions({
    int page = 1,
    int limit = 20,
    String? type,
  }) async {
    try {
      final response = await _apiService.get(
        '/customer/wallet/transactions',
        queryParameters: {
          'page': page,
          'limit': limit,
          if (type != null) 'type': type,
        },
      );

      if (response.data['success']) {
        return WalletTransactionsResponse.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل جلب المعاملات');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء جلب المعاملات',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }

  /// Initiate wallet top-up
  Future<WalletTopUpResponse> initiateTopUp({
    required double amount,
    required String paymentMethod, // 'card' or 'mobile_wallet'
    String? phoneNumber, // Required for mobile_wallet
  }) async {
    try {
      final response = await _apiService.post(
        '/customer/wallet/topup',
        data: {
          'amount': amount,
          'paymentMethod': paymentMethod,
          if (phoneNumber != null) 'phoneNumber': phoneNumber,
        },
      );

      if (response.data['success']) {
        return WalletTopUpResponse.fromJson(response.data['data']);
      } else {
        throw Exception(response.data['message'] ?? 'فشل بدء عملية الشحن');
      }
    } on DioException catch (e) {
      if (e.response != null) {
        throw Exception(
          e.response?.data['message'] ?? 'حدث خطأ أثناء بدء عملية الشحن',
        );
      } else {
        throw Exception('لا يوجد اتصال بالإنترنت');
      }
    }
  }
}

/// Wallet Balance Model
class WalletBalance {
  final double balance;
  final double totalTopups;
  final double totalSpent;

  WalletBalance({
    required this.balance,
    required this.totalTopups,
    required this.totalSpent,
  });

  factory WalletBalance.fromJson(Map<String, dynamic> json) {
    return WalletBalance(
      balance: (json['balance'] as num).toDouble(),
      totalTopups: (json['totalTopups'] as num).toDouble(),
      totalSpent: (json['totalSpent'] as num).toDouble(),
    );
  }
}

/// Wallet Transaction Model
class WalletTransaction {
  final String id;
  final String type;
  final double amount;
  final String status;
  final String? description;
  final String? orderId;
  final DateTime createdAt;

  WalletTransaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.status,
    this.description,
    this.orderId,
    required this.createdAt,
  });

  factory WalletTransaction.fromJson(Map<String, dynamic> json) {
    return WalletTransaction(
      id: json['_id'] as String,
      type: json['type'] as String,
      amount: (json['amount'] as num).toDouble(),
      status: json['status'] as String,
      description: json['description'] as String?,
      orderId: json['orderId'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  String get typeLabel {
    switch (type) {
      case 'wallet_topup':
        return 'شحن المحفظة';
      case 'order_payment':
        return 'دفع طلب';
      case 'refund':
        return 'استرجاع';
      default:
        return type;
    }
  }

  bool get isCredit => type == 'wallet_topup' || type == 'refund';
}

/// Wallet Transactions Response
class WalletTransactionsResponse {
  final List<WalletTransaction> transactions;
  final int total;
  final int page;
  final int limit;
  final int pages;

  WalletTransactionsResponse({
    required this.transactions,
    required this.total,
    required this.page,
    required this.limit,
    required this.pages,
  });

  factory WalletTransactionsResponse.fromJson(Map<String, dynamic> json) {
    // Backend sendPaginated puts transactions in 'data' field, not 'transactions'
    final List<dynamic> transactionsData =
        (json['data'] ?? json['transactions'] ?? []) as List<dynamic>;
    final pagination = json['pagination'] as Map<String, dynamic>? ?? {
      'total': 0,
      'page': 1,
      'limit': 20,
      'pages': 1,
    };

    return WalletTransactionsResponse(
      transactions: transactionsData
          .map((t) => WalletTransaction.fromJson(t as Map<String, dynamic>))
          .toList(),
      total: pagination['total'] as int? ?? 0,
      page: pagination['page'] as int? ?? 1,
      limit: pagination['limit'] as int? ?? 20,
      pages: pagination['pages'] as int? ?? 1,
    );
  }
}

/// Wallet Top-Up Response
class WalletTopUpResponse {
  final String transactionId;
  final String paymentUrl;
  final double amount;

  WalletTopUpResponse({
    required this.transactionId,
    required this.paymentUrl,
    required this.amount,
  });

  factory WalletTopUpResponse.fromJson(Map<String, dynamic> json) {
    return WalletTopUpResponse(
      transactionId: json['transactionId'] as String,
      // Backend returns 'redirectUrl', map it to paymentUrl
      paymentUrl: (json['redirectUrl'] ?? json['paymentUrl']) as String,
      amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

File: backend/src/services/wallet.service.ts
import { Types } from 'mongoose';
import { Customer, ICustomer } from '../models/Customer';
import { Transaction, ITransaction } from '../models/Transaction';
import { User } from '../models/User';
import { NotFoundError, BadRequestError } from '../utils/errors';
import { IPaginatedResult } from '../types';
import { paymobService } from './paymob.service';
import { transactionService } from './transaction.service';

interface TopupWalletInput {
  customerId: string;
  amount: number;
  paymentMethod: 'card' | 'mobile_wallet';
  phoneNumber?: string;
}

interface GetWalletTransactionsOptions {
  customerId: string;
  type?: string;
  startDate?: Date;
  endDate?: Date;
  page?: number;
  limit?: number;
}

interface WalletBalance {
  balance: number;
  totalTopups: number;
  totalSpent: number;
}

class WalletService {
  /**
   * Get customer wallet balance
   */
  async getBalance(customerId: string): Promise<WalletBalance> {
    const customer = await Customer.findOne({ userId: new Types.ObjectId(customerId) });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    return {
      balance: customer.walletBalance,
      totalTopups: customer.totalWalletTopups,
      totalSpent: customer.totalWalletSpent,
    };
  }

  /**
   * Get wallet transactions
   */
  async getTransactions(
    options: GetWalletTransactionsOptions
  ): Promise<IPaginatedResult<ITransaction>> {
    const { customerId, type, startDate, endDate, page = 1, limit = 20 } = options;

    const customer = await Customer.findOne({ userId: new Types.ObjectId(customerId) });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    const query: Record<string, unknown> = {
      $or: [
        { fromUserId: new Types.ObjectId(customerId) },
        { toUserId: new Types.ObjectId(customerId) },
      ],
      type: { $in: ['wallet_topup', 'order_payment', 'refund'] },
    };

    if (type) {
      query.type = type === 'topup' ? 'wallet_topup' : type;
    }

    if (startDate || endDate) {
      query.createdAt = {};
      if (startDate) {
        (query.createdAt as Record<string, unknown>).$gte = startDate;
      }
      if (endDate) {
        (query.createdAt as Record<string, unknown>).$lte = endDate;
      }
    }

    const skip = (page - 1) * limit;

    const [transactions, total] = await Promise.all([
      Transaction.find(query)
        .sort({ createdAt: -1 })
        .skip(skip)
        .limit(limit)
        .populate('orderId', 'orderNumber')
        .lean(),
      Transaction.countDocuments(query),
    ]);

    const pages = Math.ceil(total / limit);

    return {
      data: transactions as ITransaction[],
      pagination: {
        total,
        page,
        limit,
        pages,
        hasNext: page < pages,
        hasPrev: page > 1,
      },
    };
  }

  /**
   * Initiate wallet top-up
   */
  async initiateTopup(input: TopupWalletInput): Promise<{ redirectUrl: string; transactionId: string }> {
    const { customerId, amount, paymentMethod, phoneNumber } = input;

    // Validate customer
    const customer = await Customer.findOne({ userId: new Types.ObjectId(customerId) });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    const user = await User.findById(customerId);

    if (!user) {
      throw new NotFoundError('المستخدم غير موجود');
    }

    // Create pending transaction
    const transaction = await transactionService.createTransaction({
      type: 'wallet_topup',
      toUserId: customerId,
      amount,
      fee: 0,
      paymentMethod,
      paymentGateway: 'paymob',
    });

    // Prepare billing data
    const billingData = {
      email: user.email,
      first_name: user.name.split(' ')[0] || 'Customer',
      last_name: user.name.split(' ').slice(1).join(' ') || 'User',
      phone_number: user.phone || phoneNumber || '01000000000',
      street: 'NA',
      building: 'NA',
      floor: 'NA',
      apartment: 'NA',
      city: 'Bagour',
      country: 'EG',
      shipping_method: 'NA',
    };

    try {
      // Create payment via Paymob
      const payment = await paymobService.createWalletTopupPayment(
        transaction._id.toString(),
        amount,
        billingData,
        paymentMethod,
        phoneNumber
      );

      // Update transaction with Paymob order ID
      transaction.paymentId = payment.paymobOrderId;
      transaction.paymentData = {
        paymobOrderId: payment.paymobOrderId,
      };
      await transaction.save();

      return {
        redirectUrl: payment.redirectUrl,
        transactionId: transaction._id.toString(),
      };
    } catch (error) {
      // Mark transaction as failed
      transaction.status = 'failed';
      await transaction.save();
      throw error;
    }
  }

  /**
   * Confirm wallet top-up after payment
   */
  async confirmTopup(transactionId: string, paymentId: string): Promise<ICustomer> {
    const transaction = await Transaction.findById(transactionId);

    if (!transaction) {
      throw new NotFoundError('المعاملة غير موجودة');
    }

    if (transaction.type !== 'wallet_topup') {
      throw new BadRequestError('نوع المعاملة غير صحيح');
    }

    if (transaction.status === 'completed') {
      throw new BadRequestError('المعاملة مكتملة بالفعل');
    }

    if (!transaction.toUserId) {
      throw new BadRequestError('معرف العميل غير موجود في المعاملة');
    }

    // Update transaction
    transaction.status = 'completed';
    transaction.paymentId = paymentId;
    transaction.processedAt = new Date();
    await transaction.save();

    // Update customer wallet balance
    const customer = await Customer.findOne({ userId: transaction.toUserId });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    customer.walletBalance += transaction.amount;
    customer.totalWalletTopups += transaction.amount;
    await customer.save();

    return customer;
  }

  /**
   * Deduct from wallet for order payment
   */
  async deductFromWallet(
    customerId: string,
    orderId: string,
    amount: number
  ): Promise<{ success: boolean; newBalance: number }> {
    const customer = await Customer.findOne({ userId: new Types.ObjectId(customerId) });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    if (customer.walletBalance < amount) {
      throw new BadRequestError('رصيد المحفظة غير كافٍ');
    }

    // Create transaction
    await transactionService.createTransaction({
      type: 'order_payment',
      orderId,
      fromUserId: customerId,
      amount,
      fee: 0,
      paymentMethod: 'wallet',
    });

    // Deduct from wallet
    customer.walletBalance -= amount;
    customer.totalWalletSpent += amount;
    await customer.save();

    // Mark transaction as completed
    const transaction = await Transaction.findOne({
      orderId: new Types.ObjectId(orderId),
      type: 'order_payment',
      status: 'pending',
    }).sort({ createdAt: -1 });

    if (transaction) {
      transaction.status = 'completed';
      transaction.processedAt = new Date();
      await transaction.save();
    }

    return {
      success: true,
      newBalance: customer.walletBalance,
    };
  }

  /**
   * Refund to wallet
   */
  async refundToWallet(
    customerId: string,
    orderId: string,
    amount: number
  ): Promise<ICustomer> {
    const customer = await Customer.findOne({ userId: new Types.ObjectId(customerId) });

    if (!customer) {
      throw new NotFoundError('بيانات العميل غير موجودة');
    }

    // Create refund transaction
    const transaction = await transactionService.createTransaction({
      type: 'refund',
      orderId,
      toUserId: customerId,
      amount,
      fee: 0,
      paymentMethod: 'wallet',
      notes: 'استرجاع مبلغ الطلب إلى المحفظة',
    });

    // Add to wallet
    customer.walletBalance += amount;
    await customer.save();

    // Mark transaction as completed
    transaction.status = 'completed';
    transaction.processedAt = new Date();
    await transaction.save();

    return customer;
  }
}

export const walletService = new WalletService();
export default walletService;
